Add immutable TreeSeqMap (a SeqMap implemented via a customized IntMap/HashMap pair)

odd opened the pull request:
This is an alternative `SeqMap` implementation to compare against `VectorMap`. It uses a customized `IntMap` to provide insertion/modification iteration ordering and a `HashMap` to allow efficient lookup by key.

odd committed:
Add immutable TreeSeqMap (a SeqMap implemented via a customized IntMap/HashMap pair)

NthPortal commented:
I gave this a quick look this morning, and it looks really good. I will try to give it a full review later this week or this weekend.

cheers!

joshlemer commented:
I was thinking that since there's nowhere where you're using the TreeMap's lookup functionality, you could get by with a persistent min-max heap. Unfortunately none exist in the standard library, but maybe we could create a `private[collection]` implementation and see if it really has lower overhead.

mdedetrich commented:
Nice work, I think in general a properly optimized `VectorMap` should theoretically be faster than this `OrderedMap`, but we should benchmark the two implementations vigorously. There is also a tradeoff in terms of memory usage (which can effect performance due to cache lines as well if the collection is large enough)

odd commented:
Thanks, @mdedetrich! I must stress that this PR is highly inspired by, and partly based on, your `VectorMap` PR, but with the `Vector` switched out for a `TreeMap`. My main objective was to create a simple straightforward baseline implementation of `SeqMap` that other implementations (e.g. `VectorMap`) could be compared to and benchmarked against. There might also exist better-suited data structures than `TreeMap` to use for the ordering part in this implementation (as @joshlemer suggests above). Creating a benchmarking suite to run these against each other (both time and memory wise) feels like a good next step. Perhaps your work on the new JSON implementation can provide some relevant real-world samples to test on?

mdedetrich commented:
@odd Yup completely understand, I am curious what the real world results. Apologies if I comment came off as being too critical. The devil is usually in the detail with these things, I will also try and optimize both implementations if I have enough time

odd commented:
I pushed an updated version which replaces the `immutableTreeMap` with a customized `immutable.IntMap` leading to a 25% improvement in performance on average. I will follow up with a comparison between this implementation and `immutable.VectorMap` but I have started on some performance improvements to `immutable.VectorMap` that I would like to finish first.

odd commented:
For reference, the performance improvements to `VectorMap` I mentioned in my previous comment has been implemented in PR #7193, and while improving performance for consecutive removes (i.e. removing elements in the same order they where added) with a factor of 5 when removing 2048 consecutive elements out of a map of size 4096, `OrderedMap` is still faster than `VectorMap` by a factor of 200.

NthPortal commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,426 @@
+    final case object Modification extends OrderBy
+  }
+
+  val Empty = new OrderedMap[Nothing, Nothing](Ordering.empty, HashMap.empty, 0, OrderBy.Insertion)
```
I would recommend starting `ordinal` at `Int.MinValue`, rather than `0`. If you start at `0` and have a map containing just below `Int.MaxValue` elements, adding to and removing from that map is going to cause a rebuilding of the map every few additions, which would be extremely expensive.

Alternatively, you could use a `Long` for `ordinal`; I don't know to what degree that would impact performance or allocation size.

odd replied:
The reason that the ordinal counting starts with `0: Int`, jumps from `Int.MaxValue` to `Int.MinValue` and repopulates a fresh instance once it reaches `-1: Int` is that the `IntMap` used for the ordering uses a tree structure that is based on the bit sequence of the `Int` keys. This means that `0: Int` (`00000000000000000000000000000000`) will be the leftmost (first) possible leaf in the tree, `Int.MaxValue` (`01111111111111111111111111111111`) will be followed directly by `Int.MinValue` (`10000000000000000000000000000000`), and the rightmost (last) possible leaf is `-1: Int` (`11111111111111111111111111111111`).

```scala
scala> IntMap(1 -> "one", 10 -> "ten", Int.MinValue -> "min", -10 -> "-ten", 0 -> "zero", -1 -> "-one", Int.MaxValue -> "max")
res55: scala.collection.immutable.IntMap[String] = IntMap(0 -> zero, 1 -> one, 10 -> ten, 2147483647 -> max, -2147483648 -> min, -10 -> -ten, -1 -> -one)
```

NthPortal replied:
ah, I didn't realize that. Thanks for the explanation!

odd commented:
@SethTisue Assuming both `VectorMap` and `OrderedMap` pass the collection laws, the question becomes which has the most desired overall performance characteristics (time/memory). I have started on a benchmarking suite in odd/collection-benchmark which indicate similar performance (time) between them for all operations except removal of sequential elements (see https://github.com/scala/scala/pull/7146#issuecomment-427575921) where `OrderedMap` is significantly faster. I plan on adding more extensive benchmarking once the lawfullness of both has been be verified.

mdedetrich commented:
@SethTisue Afaik the issue is its hard to say which collection is overall better. `OrderedMap` is better if you remove keys very often, where as if this isn't the case than `VectorMap` is better overall.

Then there is also memory to to take into account. @odd Is there a list of benchmarks somewhere which show how much faster `VectorMap` is compared to `OrderedMap` on common operations (i.e. lookup by key, iteration, etc etc)?

joshlemer commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,426 @@
+      result = result.updated(k, v)
+    }
+    result
+  }
```
if `suffix` is a mutable/immutable HashMap, `concat` could avoid rehashing keys when inserting into the OrderedMap's HashMap.

Same with `removedAll` if the argument is a mutable/immutable HashSet

odd replied:
@joshlemer I'm a bit uncertain as to how that would be accomplished so I'll take a look at the implementations in HashMap/HashSet for guidance.

joshlemer replied:
This is the PR to check https://github.com/scala/scala/pull/7520. Maybe it's easier to just wait until that is merged so that you get the private[collection] methods `immutable.HashMap/HashSet#foreachWithHash` and `mutable.HashMap/HashSet#nodeIterator`

specifically, 

https://github.com/scala/scala/pull/7520/files#diff-5be35908b1ec9bbb90c66342b71dd9afR90

shows how both can be used.

odd replied:
@joshlemer Do you still feel that this issue needs addressing considering the latest set of changes (00b7f1e)?

joshlemer replied:
Very cool! 👏 💐

joshlemer replied:
Though, this particular comment is more about extracting the hashes out of `suffix` in the case that it happens to be a `HashMap` That way, you can insert the keys into your builder without re-hashing. 

This could be considered a stretch goal though, I don't think it needs to block this PR, maybe we should make a separate backlog-milestone issue for it.

joshlemer commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,426 @@
+  def newBuilder[K, V](orderBy: OrderBy): mutable.Builder[(K, V), OrderedMap[K, V]] =
+    new mutable.ImmutableBuilder[(K, V), OrderedMap[K, V]](empty(orderBy)) {
+      def addOne(elem: (K, V)): this.type = { elems = elems + elem; this }
+    }
```
Could create a mutating builder for much better performance.

odd replied:
Same as above, I'll look at HashMap for guidance.

odd replied:
Addressed by 00b7f1e.

odd commented:
I have now (finally) updated and pushed my latest benchmarking code to [collection-benchmark](http://github.com/odd/collection-benchmark); here are the results comparing the three `immutable.SeqMap` implementations `ListMap`, `VectorMap` and `OrderedMap`:
* Time: http://jmh.morethan.io/?gist=5f104258be4a62f850ca71a9397cd527 (best viewed using logarithmic scale)
* Memory: http://jmh.morethan.io/?gist=a9e7b197451a714f77b6e75e7c495d67 (best viewed using linear scale)

mdedetrich commented:
@odd Honestly I am personally borderline, completely understand (and agree!) with your points, the issue I have is that `VectorMap` is already a memory intensive collection (its actually what I would consider borderline in regards to amortized memory used vs collection size) and `OrderedMap` just makes this much worse.

Its the whole using roughly 50% more memory than a collection which is already borderline in terms of acceptable memory usage that I have an issue with (and collections that use a lot of memory pose their own problems, with such high memory churn you are often competing with cache lines versus other resources which isn't something that shows up in benchmarks. Compounded ontop of this we are dealing with the JVM environment which doesn't have the best track record when it comes to memory conservation).

So to me the whole equation is whether we should balance using a **huge** amount of memory against an operation which is at least acceptable? in terms of performance.

@Ichoran @NthPortal @joshlemer @julienrf @NthPortal What are your thoughts on this?

Ichoran commented:
I'm inclined to agree with @odd that in their current incarnations, `OrderedMap` is a better default collection than `VectorMap` for most use cases.  Despite the substantial per-element overhead (~150 bytes vs. 100 for `VectorMap`), it's considerably faster at many common operations (e.g. filter, map, collect) without being considerably slower in anything.  That seems, to me, to be the more desirable default behavior: pretty good at everything with no serious deficits.

On the other hand, I'm a little surprised at the results for `VectorMap`, so it may be that `VectorMap` can be sped up substantially.  In that case, if you ask me which architecture seems more promising, Vector + Hash vs. IntHash + Hash, I'd think the former.  While `Vector` isn't great at removals, it's also not that common for immutable collections to have a lot of single-element removals, and in other cases the `Vector` architecture _ought_ to be better than the custom `IntMap`.  Otherwise, why don't we throw away `Vector` and replace it with an `IntMap`?

@mdedetrich, do you have time to track down why `VectorMap` seems slow with operations like `map`?

Ichoran commented:
Also @odd how do you know that `OrderedMap` is correct?  There are, of course, a lot of ways to go faster if you're not constrained to produce the right answer :)

(I've run collections-laws on `VectorMap` and it passes all the tests I've got.  That said, there aren't extensive tests for the in-order property on maps.)

joshlemer commented:
@Ichoran if you're talking about the variant of `map` that takes an `((K, V)) => (X, Y)` and returns `VectorMap[X, Y]`, it's probably slow because VectorMap's builder is very slow (see https://github.com/scala/scala/pull/7508).

odd commented:
@Ichoran I've been meaning to run `OrderedMap` through collections-laws and make sure it passes, but I haven't been able to do so yet (last I checked @SethTisue was going to make a new release of collection-laws; perhaps that has already been done, I'll check tomorrow).

mdedetrich commented:
> On the other hand, I'm a little surprised at the results for VectorMap, so it may be that VectorMap can be sped up substantially. In that case, if you ask me which architecture seems more promising, Vector + Hash vs. IntHash + Hash, I'd think the former

Yeah I mentioned that earlier https://github.com/scala/scala/pull/7146#issuecomment-417046622

> While Vector isn't great at removals, it's also not that common for immutable collections to have a lot of single-element removals

Agreed.

Im currently on holidays which will finish in a couple of weeks, so I can look at after that. What @joshlemer said could seem like a good candidate as to why `VectorMap` is slower than expected seeing as we use `Builder` almost everywhere where we can. Would be interesting to run @odd 's benchmarks after the PR is merged.

I think it also makes sense to revisit this topic when a lot of these performance improvement PR's are merged seeing as not even the original collections (which these newer collections rely on) still need performance improvements.

Ichoran commented:
@odd - Do the instructions I have on the collections-laws readme not work for you?  There are a few steps, but you should be able to get the tests running without `OrderedMap` without having to figure out anything complicated.

The docs also contain an example of adding a new collection.  So that should also be reasonably straightforward (though I might need to update the example to work flawlessly with 2.13 instead of 2.12 collections).

Alternatively, I could try it out when I get time.

odd commented:
@Ichoran I did try to run collection-laws a while ago but did not manage to get sourcecode to publish locally. I'll give it another try and see if I fare better this time.

Ichoran commented:
@odd - I have instructions in the README about how to change sourcecode to get it compiling in a minimal state needed for collections-laws.

odd commented:
@Ichoran Those instructions is what I have been following but still I can not seem to get sourcecode published locally for my local pre-snapshot version of Scala. It fails with the following exception:
```
[error] java.lang.NullPointerException
[error] 	at scala.tools.nsc.Driver.scalacError(Driver.scala:18)
[error] 	at scala.tools.nsc.Driver.$anonfun$process$1(Driver.scala:41)
[error] 	at scala.tools.nsc.Driver.$anonfun$process$1$adapted(Driver.scala:41)
[error] 	at scala.reflect.internal.util.ScalaClassLoader.error$1(ScalaClassLoader.scala:60)
[error] 	at scala.reflect.internal.util.ScalaClassLoader.create(ScalaClassLoader.scala:80)
[error] 	at scala.reflect.internal.util.ScalaClassLoader.create$(ScalaClassLoader.scala:58)
[error] 	at scala.tools.nsc.reporters.Reporter$$anon$2.create(Reporter.scala:87)
[error] 	at scala.tools.nsc.reporters.Reporter$.apply(Reporter.scala:88)
[error] 	at scala.tools.nsc.Driver.process(Driver.scala:42)
[error] 	at scala.tools.nsc.Main.process(Main.scala)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.lang.reflect.Method.invoke(Method.java:498)
[error] 	at sbt.internal.inc.RawCompiler.getReporter$1(RawCompiler.scala:50)
[error] 	at sbt.internal.inc.RawCompiler.apply(RawCompiler.scala:71)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$6(AnalyzingCompiler.scala:350)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.handleCompilationError$1(AnalyzingCompiler.scala:327)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4(AnalyzingCompiler.scala:346)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$4$adapted(AnalyzingCompiler.scala:341)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:376)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:383)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$2(AnalyzingCompiler.scala:341)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.$anonfun$compileSources$2$adapted(AnalyzingCompiler.scala:335)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:376)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:383)
[error] 	at sbt.internal.inc.AnalyzingCompiler$.compileSources(AnalyzingCompiler.scala:335)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$3(ZincComponentCompiler.scala:277)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$3$adapted(ZincComponentCompiler.scala:262)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:376)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:383)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$2(ZincComponentCompiler.scala:262)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] 	at sbt.internal.util.BufferedLogger.bufferQuietly(BufferedLogger.scala:106)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$1(ZincComponentCompiler.scala:262)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compileAndInstall$1$adapted(ZincComponentCompiler.scala:259)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:376)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:383)
[error] 	at sbt.internal.inc.ZincComponentCompiler.compileAndInstall(ZincComponentCompiler.scala:259)
[error] 	at sbt.internal.inc.ZincComponentCompiler.$anonfun$compiledBridgeJar$1(ZincComponentCompiler.scala:225)
[error] 	at sbt.internal.inc.IfMissing$Define.run(IfMissing.scala:19)
[error] 	at sbt.internal.inc.ZincComponentManager.createAndCache$1(ZincComponentManager.scala:46)
[error] 	at sbt.internal.inc.ZincComponentManager.$anonfun$files$3(ZincComponentManager.scala:57)
[error] 	at sbt.internal.inc.ZincComponentManager.getOrElse$1(ZincComponentManager.scala:38)
[error] 	at sbt.internal.inc.ZincComponentManager.$anonfun$files$2(ZincComponentManager.scala:57)
[error] 	at sbt.internal.inc.ZincComponentManager$$anon$1.call(ZincComponentManager.scala:87)
[error] 	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:95)
[error] 	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:80)
[error] 	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:99)
[error] 	at xsbt.boot.Using$.withResource(Using.scala:10)
[error] 	at xsbt.boot.Using$.apply(Using.scala:9)
[error] 	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:60)
[error] 	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:50)
[error] 	at xsbt.boot.Locks$.apply0(Locks.scala:31)
[error] 	at xsbt.boot.Locks$.apply(Locks.scala:28)
[error] 	at sbt.internal.inc.ZincComponentManager.lock(ZincComponentManager.scala:87)
[error] 	at sbt.internal.inc.ZincComponentManager.$anonfun$lockSecondaryCache$1(ZincComponentManager.scala:84)
[error] 	at scala.Option.map(Option.scala:146)
[error] 	at sbt.internal.inc.ZincComponentManager.lockSecondaryCache(ZincComponentManager.scala:82)
[error] 	at sbt.internal.inc.ZincComponentManager.fromSecondary$1(ZincComponentManager.scala:55)
[error] 	at sbt.internal.inc.ZincComponentManager.$anonfun$files$6(ZincComponentManager.scala:61)
[error] 	at sbt.internal.inc.ZincComponentManager.getOrElse$1(ZincComponentManager.scala:38)
[error] 	at sbt.internal.inc.ZincComponentManager.$anonfun$files$5(ZincComponentManager.scala:61)
[error] 	at sbt.internal.inc.ZincComponentManager$$anon$1.call(ZincComponentManager.scala:87)
[error] 	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:95)
[error] 	at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:80)
[error] 	at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:99)
[error] 	at xsbt.boot.Using$.withResource(Using.scala:10)
[error] 	at xsbt.boot.Using$.apply(Using.scala:9)
[error] 	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:60)
[error] 	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:50)
[error] 	at xsbt.boot.Locks$.apply0(Locks.scala:31)
[error] 	at xsbt.boot.Locks$.apply(Locks.scala:28)
[error] 	at sbt.internal.inc.ZincComponentManager.lock(ZincComponentManager.scala:87)
[error] 	at sbt.internal.inc.ZincComponentManager.lockLocalCache(ZincComponentManager.scala:78)
[error] 	at sbt.internal.inc.ZincComponentManager.files(ZincComponentManager.scala:61)
[error] 	at sbt.internal.inc.ZincComponentManager.file(ZincComponentManager.scala:66)
[error] 	at sbt.internal.inc.ZincComponentCompiler.compiledBridgeJar(ZincComponentCompiler.scala:225)
[error] 	at sbt.internal.inc.ZincComponentCompiler$ZincCompilerBridgeProvider.compiledBridge(ZincComponentCompiler.scala:83)
[error] 	at sbt.internal.inc.ZincComponentCompiler$ZincCompilerBridgeProvider.fetchCompiledBridge(ZincComponentCompiler.scala:90)
[error] 	at sbt.internal.inc.AnalyzingCompiler.loader(AnalyzingCompiler.scala:249)
[error] 	at sbt.internal.inc.AnalyzingCompiler.getInterfaceClass(AnalyzingCompiler.scala:267)
[error] 	at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:234)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:166)
[error] 	at sbt.internal.inc.AnalyzingCompiler.doc(AnalyzingCompiler.scala:147)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1(Doc.scala:39)
[error] 	at sbt.Doc$.$anonfun$scaladoc$1$adapted(Doc.scala:39)
[error] 	at sbt.RawCompileLike$.$anonfun$prepare$1(RawCompileLike.scala:83)
[error] 	at sbt.RawCompileLike$.$anonfun$prepare$1$adapted(RawCompileLike.scala:76)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$3(RawCompileLike.scala:67)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$3$adapted(RawCompileLike.scala:65)
[error] 	at sbt.util.Tracked$.$anonfun$inputChanged$1(Tracked.scala:149)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$1(RawCompileLike.scala:72)
[error] 	at sbt.RawCompileLike$.$anonfun$cached$1$adapted(RawCompileLike.scala:55)
[error] 	at sbt.Defaults$.$anonfun$docTaskSettings$3(Defaults.scala:1422)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:40)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:67)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:269)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:278)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:269)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
```

odd commented:
@Ichoran It seems I forgot to publish the compiler and reflect sub-projects for my local Scala 2.13 snapshot build. With those published locally sourcecode built successfully on 2.13.

odd commented:
@Ichoran I have now successfully run collections-laws for `OrderedMap` (one issue was found concerning concatenation, it has been fixed in a4674d0). 
While running the laws I noticed a lot of errors for `immutable.LazyList` and some for `immutable.BitSet` but perhaps those are already known.

Ichoran commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,518 @@
+  final object OrderBy {
+    final case object Insertion extends OrderBy
+    final case object Modification extends OrderBy
+  }
```
This is an interesting feature, but I'm not sure the complexity pays for itself.  Insertion-order preservation is by far the most common since in most implementations it's more efficient.  If you want ordering by modification, the thing to do is remove the element and then add it again.  Yes, you lose a bit of efficiency.  Anyway, that's what SeqMap does, right?  We don't want behavior to differ; it's too surprising.

odd replied:
While a user can use the remove-insert method for explicit calls to get modification order, as it is now the ordering also affects implicit reinsertions (such as those that occur during transformation calls (`map`/`collect`/`flatMap`) when the transformation function returns identical keys for different entries. For this reason I have kept the `orderingBy`-feature for now but it would be easy to remove it if so required).

Ichoran commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,518 @@
+        val (front, rest) = this.ordering.splitAt(f)
+        val (ordering, rear) = rest.splitAt(u - f)
+        var mapping = this.mapping
+        val frontIter = front.iterator
```
You should keep track of the sizes and rebuild the mapping from scratch if the slice is small.  Not sure exactly what the threshold should be; I think removals are more expensive than additions.

odd replied:
I have added support for this (if the slice will contain more than half the original entries then the above peel-off-front-and-rear approach is used, otherwise a builder-based approach is used instead). More benchmarking should probably be done to get a more exact pivot-point.

Ichoran commented on src/library/scala/collection/immutable/OrderedMap.scala:
```diff
@@ -0,0 +1,518 @@
+    mapping.get(key) match {
+      case e if ordinal == -1 && (orderedBy == OrderBy.Modification || e.isEmpty) ⇒
+        // Reinsert into fresh instance to restart ordinal counting, expensive but only done after 2^32 updates.
+        OrderedMap.empty[K, V](orderedBy) ++ this + (key → value)
```
Please write out arrows as `->`. People can use ligature fonts if they want to view them in a cleaner shape; unicode arrows are hard for most people to type (not set up to do it).

He-Pin replied:
And it's on the way to be deleted ？

SethTisue replied:
indeed

odd replied:
Fixed in commit 00b7f1e.

Ichoran reviewed (changes requested):
This is an impressive bit of work!  I have one request for a simplification and some other minor comments, but it looks pretty good.

One bikeshedding thing, though--I don't think `OrderedMap` is obviously the name of a `SeqMap` instead of a `SortedMap`.  If you want a similar name, I think `InOrderMap` might be just non-Sorted-sounding enough to work.  Alternatively, maybe you can come up with something else?

mdedetrich commented:
@odd Do you want to rerun your benchmarks now that https://github.com/scala/scala/pull/7508 is merged?

odd commented:
@mdedetrich I will rerun the benchmarks later today on a local Scala build which includes #7508 and also an updated `OrderedMap`-implementation that uses the same techniques as in #7508 for the mapping and combines this with an in-place append operation for the ordering tree (see 00b7f1e). The benchmark will compare `OrderedMap` (the previous version), `OrderedMapY` (the latest version) and `VectorMap` (the latest version).

odd commented:
@Ichoran Regarding the name `OrderedMap`, I agree that it could be somewhat confusing (especially since `SortedMap` uses an `Ordering` to sort its elements). Some suggestions for alternative names are: 
* `SeqHashMap`
* `LinkedMap`/`LinkedHashMap` (although it is not technically _linked_ like the mutable one)

odd commented:
@joshlemer I considered `TreeHashMap` but since `immutable.HashMap` is also implemented as a tree, it was a bit undescriptive. What do you think about `SeqHashMap`?

joshlemer commented:
Hmm, well I don't like it a lot because there isn't actually a Seq anywhere backing it, like a `Vector` backs a `VectorMap` or a `List` backs a `ListMap`. Maybe a `TreeSeqMap`?

odd commented:
@joshlemer But the principal interface is called `SeqMap` and it isn't a `Seq` (as in `collection.Seq`) either, so `SeqHashMap` would be short for `Sequential Hash Map` pertaining to the entries in the map being _sequential_ (as opposed to the meaning of `Seq` in `collection.Seq` where `Seq` stands for _sequence_; whether there's a meaningful difference between the terms I don't know - I'm not a native English speaker so the nuances might be lost on me I'm afraid).

joshlemer commented:
I think it makes most sense to call it a TreeSeqMap because what it is is a SeqMap (not a HashMap, though it contains one that is not what it is). It's a SeqMap which is backed by both a HashMap and a Tree, so I think the last word should be SeqMap, and then at the front put something like `TreeSeqMap`/`TrieSeqMap`/`TreeHashSeqMap`/`TrieHashSeqMap`.

I think the `Tree`/`Trie` aspect are more important to differentiate because remember that VectorMap also uses a HashMap, so SeqHashMap could equally apply to VectorMap. Anyways, just my $0.02 😄

Ichoran reviewed (changes requested):
Just need to rename to `TreeSeqMap` and then I guess it's okay.  I still am uneasy about the ordering policy, but I will leave that to someone else to decide.

odd commented:
@Ichoran I have changed the name and squashed and pushed but it still looks like there is change requested by you. The problem is that I cannot find your comments any more (I had to do a force push to avoid the earlier commit that failed by Codacy so the change request might have got sidelined somehow). Please let me know if you think anything more should be changed (besides your reservation about the modification ordering).

Ichoran commented:
Looks good!  (Still uneasy about the insertion/modification thing, but not so uneasy that we should hold this up.)

joshlemer commented on src/library/scala/collection/immutable/TreeSeqMap.scala:
```diff
@@ -0,0 +1,638 @@
+    final case object Modification extends OrderBy
+  }
+
+  val Empty = new TreeSeqMap[Nothing, Nothing](Ordering.empty, HashMap.empty, 0, OrderBy.Insertion)
```
Would it make sense to start at `Int.MinValue` rather than `0`?

joshlemer replied:
(Sorry for the nitpick)

Ichoran replied:
Positive indexing is easier.  I'd leave it alone.

odd replied:
@joshlemer @SethTisue I'll repeat the answer I gave to @NthPortal (https://github.com/scala/scala/pull/7146#discussion_r226823340) when he asked the same question: 

> The reason that the ordinal counting starts with `0: Int`, jumps from `Int.MaxValue` to `Int.MinValue` and repopulates a fresh instance once it reaches `-1: Int` is that the `IntMap` used for the ordering uses a tree structure that is based on the bit sequence of the `Int` keys. This means that `0: Int` (00000000000000000000000000000000) will be the leftmost (first) possible leaf in the tree, `Int.MaxValue` (01111111111111111111111111111111) will be followed directly by `Int.MinValue` (10000000000000000000000000000000), and the rightmost (last) possible leaf is `-1: Int` (11111111111111111111111111111111).
```scala
scala> IntMap(1 -> "one", 10 -> "ten", Int.MinValue -> "min", -10 -> "-ten", 0 -> "zero", -1 -> "-one", Int.MaxValue -> "max")
res55: scala.collection.immutable.IntMap[String] = IntMap(0 -> zero, 1 -> one, 10 -> ten, 2147483647 -> max, -2147483648 -> min, -10 -> -ten, -1 -> -one)
```

mdedetrich commented:
@odd Thanks for the benchmarks, there are definitely substantial improvements in performance to `VectorMap`. I think there are still areas where we can squeeze more performance, will check this out on the weekend.

odd commented:
/rebuild

SethTisue commented:
#8041 proposes to change the default to `TreeSeqMap`
